### PR TITLE
Add nRF Cloud MQTT A-GPS filtered support

### PR DESCRIPTION
**Pull request by plskeggs**
Add support for filtered ephemerides:
- encode new parameters in JSON assistance request
- set GNSS elevation threshold to match configured value for filtered mode, and suppress scheduled downloads, in ATv2, location library, and nrf9160_gps driver (so ATv1 gets this too) -- but only for the apps and samples which use the GNSS periodic fix mode
- suppress redundant ephemerides assistance in filtered mode (limit to once every 2 hours)

To do: ensure ephemerides validity period is considered when deciding to suppress ephemerides assistance.

### CODE REVIEW

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps
Comment (hiltunent) [inline subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps:19]: How does the cloud handle this - does it send ephemerides for the satellites above the threshold at the time of request or does it send ephemerides for all the satellites that will rise above the threshold during the validity period of the ephemeris to be provided?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c
Comment (simensrostad) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:107]: In the AWS and Azure integrations in ATv2 we have our own encoding function for A-GPS requests, `cloud_codec_encode_agps_request`.

This functionality that seems to impact the A-GPS request format needs to be implemented in those cases as well. But, I don't think it makes sense to add it inside `cloud_codec_encode_agps_request`, since it is purely an encoding function.

A good place could be just before encoding in:

https://github.com/nrfconnect/sdk-nrf/blob/cfedfdfa08567b2252b511a4d1db15fbeba8152d/applications/asset_tracker_v2/src/modules/data_module.c#L673

Then you could pass in some flags/metadata into `cloud_codec_encode_agps_request` via `cloud_data_agps_request` that impacts the format of the message.

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps:19]: Good Question.  Cole has asked RX Network for details.  We'll let you know.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:107]: The additional parameters are optional.  Requests without them are still supported.

That being said, yes, I can add additional entries to the `cloud_data_agps_request` structure.

Seems to me that `cloud_codec_encode_agps_request` is more of a wrapper around another function that actually does the encoding, `json_common_agps_request_data_add()`.  Who is going to make the changes there to encode it so your cloud sides are happy?

File: applications/asset_tracker_v2/src/modules/gps_module.c
Comment (plskeggs) [reply applications/asset_tracker_v2/src/modules/gps_module.c:459]: sure

File: drivers/gps/nrf9160_gps/nrf9160_gps.c
Comment (jayteemo) [inline drivers/gps/nrf9160_gps/nrf9160_gps.c:784]: should this be `|=` ?

File: lib/location/method_gnss.c
Comment (jayteemo) [inline lib/location/method_gnss.c:619]: could this chunk of code be moved into a common function?  it is used 3 times in this PR

File: drivers/gps/nrf9160_gps/nrf9160_gps.c
Comment (plskeggs) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:784]: No.  Disabling scheduled downloads is only supported in mfw_1.3.1 or greater.  So if we fail to set it, this tries again without it.  It would be better though to mask off `NRF_MODEM_GNSS_USE_CASE_SCHED_DOWNLOAD_DISABLE` instead.

Comment (jayteemo) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:784]: ah sorry i meant `use_case &= ~NRF_MODEM_GNSS_USE_CASE_SCHED_DOWNLOAD_DISABLE;` like is done below, since the  use_case comes from `drv_data->config.use_case`

File: lib/location/method_gnss.c
Comment (plskeggs) [reply lib/location/method_gnss.c:619]: I thought about that, but it's not clear to me where that would live, since it's not nRF Cloud-specific.  Any suggestions?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:107]: I made the changes.  Let me know if you will be making the actual encoding changes.

File: lib/location/method_gnss.c
Comment (plskeggs) [reply lib/location/method_gnss.c:619]: done

File: drivers/gps/nrf9160_gps/nrf9160_gps.c
Comment (plskeggs) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c]: done

File: applications/asset_tracker_v2/src/modules/gps_module.c
Comment (simensrostad) [inline applications/asset_tracker_v2/src/modules/gps_module.c:460]:
```suggestion
	/* prevent modem from wasting time downloading missing ephemerides */
	use_case |= NRF_MODEM_GNSS_USE_CASE_SCHED_DOWNLOAD_DISABLE;
	err = nrf_modem_gnss_use_case_set(use_case);

	if (!err) {
		LOG_DBG("Disabled scheduled GNSS downloads");
	} else {
		LOG_DBG("Could not disable scheduled GNSS downloads: %d", err);
		use_case &= ~NRF_MODEM_GNSS_USE_CASE_SCHED_DOWNLOAD_DISABLE;
		err = nrf_modem_gnss_use_case_set(use_case);
		if (err) {
			LOG_ERR("Failed to set GPS use case, error: %d", err);
			return -EINVAL;
		}
	}

	err = nrf_modem_gnss_elevation_threshold_set(mask_angle);
	if (err) {
		LOG_ERR("Failed to set elevation threshold to %d: %d",
			mask_angle, err);
	}
```

Comment (simensrostad) [inline applications/asset_tracker_v2/src/modules/gps_module.c:460]: Could you make a separate static function in `gps_module.c` that replaces `nrf_cloud_agps_configure_filter`?
So that we avoid a dependency to `nrf_cloud_agps` ?

There is a dependency to `nrf_cloud_pgps` in this module, but we intend to remove that.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c
Comment (simensrostad) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:107]: I can update the encoding after this is merged, @coderbyheart needs to be in the loop. I've made a ticket:

https://projecttools.nordicsemi.no/jira/browse/IRIS-2637

**Review by jayteemo — APPROVED** [review]

File: applications/asset_tracker_v2/src/modules/gps_module.c
Comment (plskeggs) [reply applications/asset_tracker_v2/src/modules/gps_module.c:460]: No.  The whole point of making `nrf_cloud_agps_configure_filter` was to eliminate redundancy, as Justin pointed out in his review comments.

I could consider moving it to some other library, I suppose.  It's not directly related to A-GPS, but it is a strange enough function that it is not useful for many other purposes.

Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/gps_module.c:460]: Ok. Then it doesn't make much sense to have
CONFIG_GPS_MODULE_AGPS_FILTERED and CONFIG_GPS_MODULE_ELEVATION_MASK. If this functionality is especially tied up against nRF Cloud A-GPS.

File: doc/nrf/releases/release-notes-changelog.rst
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:104]: Maybe instead of imperative: ..., A-GPS assistance requests to cloud are limited to...

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps
Comment (coleHafner) [reply subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps:19]: Heard back today:

```
The filtering functionality is implemented to return the satellites above the elevation mask at the time the server processes the request.

It’s designed to ensure that the satellites received are able to be all usable for positioning at that point in time.
```

I have a question pending about how long the ephems are valid for, but haven't heard back yet. I suspect it's not fixed and based on how far over the satellites are already.

**Review by peknis — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps
Comment (hiltunent) [reply subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps:19]: Thanks. If our fix interval is long enough to trigger an A-GPS data request every time, this does not make any different - we have all the info we need for now. But if we attempt to get another fix using the previously downloaded filtered ephemeris it is likely that new satellites have appeared on the sky for which we don't have ephemeris and old satellites have disappeared from view. This may lead to a PVT solution with less satellites or even no fix at all. Another issue is that if GNSS is missing ephemeris for more than two or so SVs, it keeps requesting for A-GPS data. Filtering A-GPS requests pretty much guarantees that the modem will request assistance every time it is allowed to do so, i.e., every 60 minutes. Taking all this into account, I think we should state very clearly in the documentation that filtering A-GPS requests should only be used when the fix interval is longer than the ephemeris validity time, i.e., we need to do A-GPS request every time we make a fix anyway. Any savings in energy or data consumption due to filtering ephemeris will be lost many times over if we need to make a second request to download something we could have downloaded the first time but we filtered it out. This is mainly related to relatively large energy overhead when setting up the LTE connection.

**Review by simensrostad — APPROVED** [review]

**Review by jtguggedal — COMMENTED** [review]
Looks good to me. As described by @hiltunent, we need a note/warning in the docs that the returned, filtered ephemeris is for the satellites that are in view at the time of the request

File: include/net/nrf_cloud_agps.h
Comment (tokangas) [inline include/net/nrf_cloud_agps.h:77]: I understand that the same code is used in multiple places and it would be nice to have it in one place, but I don't think this belongs in the nRF Cloud library. I find it odd that a nRF Cloud library function is used to configure GNSS. Also, the actual functionality may not be what we really want.

It makes sense to disable scheduled downloads when A-GPS is used. However, it's not really related to the use of filtered ephemerides. The main reason for disabling scheduled downloads are almanacs, not ephemerides. Actually, GNSS only does scheduled downloads for ephemerides when the used fix interval is 30 minutes or less. And as mentioned by @hiltunent in a previous comment, it does not make sense to use filtered ephemerides with this short intervals.

It's also not mandatory to set the GNSS elevation threshold angle to the same value which is used for filtered ephemerides. It makes no sense to use for example 15 degrees for GNSS and 5 degrees for filtered ephemerides, because ephemerides for satellites between 5 and 15 degrees would be downloaded, but those would not be used by GNSS. The other way around it makes more sense, you could download ephemerides for satellites above 15 degrees, but leave the GNSS threshold at 5 degrees. In this case GNSS would still be able to use satellites between 5 and 15 degrees by downloading the missing ephemerides from the broadcast, if satellites above 15 degrees are not enough for a fix.

All in all, my opinion is that this function should be removed. It would be good to disable scheduled downloads when A-GPS is used, but maybe there should be a separate PR for that instead. And I'm not sure if it's necessary to change the GNSS elevation threshold at all. It can be different from NRF_CLOUD_AGPS_ELEVATION_MASK and in most cases there's no need to change the default value of 5 deg.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c
Comment (tokangas) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:183]:
```suggestion
	LOG_INF("Requesting filtered ephemerides with elevation mask angle = %u degrees",
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c
Comment (tokangas) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:634]:
```suggestion
	/**
	 * In filtered mode, because fewer than the full set of ephemerides is sent to
	 * the modem, determine here if we correctly received them from the cloud and
	 * sent them to the modem.
	 */
```

**Review by tokangas — CHANGES_REQUESTED** [review]

File: include/net/nrf_cloud_agps.h
Comment (plskeggs) [reply include/net/nrf_cloud_agps.h:77]: It would not make sense to filter to 15 degrees but leave the GNSS threshold to 5 -- the whole point of enabling A-GPS is to eliminate the need to wait for the GNSS unit to download ephemerides from the broadcast.  I think we discussed this elsewhere (in a call?) and it was decided that the two values should match.

Yes, I agree it doesn't belong here really (see my comment earlier: https://github.com/nrfconnect/sdk-nrf/pull/6388#discussion_r769011704).  But there is no other place I could find in the sdk-nrf repo that has any appropriate place to put such a general GNSS function.  I think there should be one that helps set the use case based on whether the NRF_MODEM_GNSS_USE_CASE_SCHED_DOWNLOAD_DISABLE is supported or not, or at least, to determine if the modem firmware present supports a given feature.  Is there such a facility?

Comment (tokangas) [reply include/net/nrf_cloud_agps.h:77]: Usually, there should be no need to change the elevation threshold from the default value. I think we at least discussed that the default for filtered ephemeris download should be the same which is used by the modem (5 deg). Having it configurable in the library is fine, but I don't think there's need for a function which configures the GNSS elevation threshold. Changing the elevation threshold with the GNSS API is a single function call with only one parameter.

There's no helper for setting the use case bitmask. There could be one for the reason you mentioned, but we don't have a good place for it, I think. However, I don't see this as a big problem, because if the modem FW doesn't support `NRF_MODEM_GNSS_USE_CASE_SCHED_DOWNLOAD_DISABLE` and setting the use case bitmask fails, in most cases the only other set bit is `NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START` and this is already enabled by default. So there's no need to call `nrf_modem_gnss_use_case_set(NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START)` in the failure case. Actually, the `NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START` bit doesn't (currently) affect GNSS behavior at all, it just has been present in the API from the start and although we considered removing it, it could have caused some problems.

Comment (plskeggs) [reply include/net/nrf_cloud_agps.h]: I've tried to accommodate most of your suggestions.  Please review again and approve if possible.

File: applications/asset_tracker_v2/src/modules/gnss_module.c
Comment (simensrostad) [inline applications/asset_tracker_v2/src/modules/gnss_module.c:463]:
```suggestion
		LOG_ERR("Failed to set elevation threshold to %d, error: %d",
```

File: applications/asset_tracker_v2/src/modules/gnss_module.c
Comment (simensrostad) [inline applications/asset_tracker_v2/src/modules/gnss_module.c:466]:
```suggestion
		LOG_DBG("Elevation threshold set to %d", CONFIG_GNSS_MODULE_ELEVATION_MASK);
```

File: applications/asset_tracker_v2/src/modules/gnss_module.c
Comment (simensrostad) [inline applications/asset_tracker_v2/src/modules/gnss_module.c:473]:
```suggestion
		/* Prevent the modem from wasting time downloading missing ephemerides */
```

File: applications/asset_tracker_v2/src/modules/gnss_module.c
Comment (simensrostad) [inline applications/asset_tracker_v2/src/modules/gnss_module.c:17]: Is this include still needed?

**Review by simensrostad — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps
Comment (tokangas) [inline subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps:39]: This option only controls the data download, GNSS needs to be configured by the application.
```suggestion
	  This sets the threshold elevation angle in A-GPS cloud requests.
	  It constrains which satellite ephemerides are included in the
	  assistance data returned by the cloud.
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c
Comment (tokangas) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_agps.c:112]:
```suggestion
	/**
	 * Determine if we processed ephemerides assistance less than 2 hours ago; if so,
	 * we can skip this.
	 */
```

File: applications/asset_tracker_v2/src/modules/gnss_module.c
Comment (tokangas) [inline applications/asset_tracker_v2/src/modules/gnss_module.c:489]: I just now realized, that this isn't actually relevant with ATv2. ATv2 uses GNSS in single fix mode and scheduled downloads are only used in periodic mode. The use case configuration here can just be removed.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c
Comment (tokangas) [inline drivers/gps/nrf9160_gps/nrf9160_gps.c:22]: This include is probably not needed. But the whole GPS driver has been deprecated since v1.7.0 and we're just discussing removing it for v1.9.0. So it's probably not necessary to update the driver anymore.

File: lib/location/method_gnss.c
Comment (tokangas) [inline lib/location/method_gnss.c:647]: Location library uses GNSS in continuous tracking mode, so scheduled downloads do not apply. These changes can be reverted.

File: doc/nrf/libraries/networking/nrf_cloud_agps.rst
Comment (tokangas) [inline doc/nrf/libraries/networking/nrf_cloud_agps.rst:48]: Better use 'LTE' to cover both M1 and NB-IoT.
```suggestion
When the application only requires a fast GNSS fix at most once per 2 hour period, it can reduce LTE data charges by enabling :kconfig:`CONFIG_NRF_CLOUD_AGPS_FILTERED`.
```

**Review by tokangas — CHANGES_REQUESTED** [review]

File: applications/asset_tracker_v2/src/modules/gnss_module.c
Comment (plskeggs) [reply applications/asset_tracker_v2/src/modules/gnss_module.c]: OK, if there are no plans to support periodic mode in ATv2 (@simensrostad ?)  I'll remove it.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c
Comment (plskeggs) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:22]: Noted.  As the saying goes, "I'll believe it when I see it." :-)

File: lib/location/method_gnss.c
Comment (plskeggs) [reply lib/location/method_gnss.c:647]: Is this always going to be the case, @hiltunent ?  Any reason to leave this in case other tracking modes are supported in the future?

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps:14]: Just below in this file, I select it on when NRF_CLOUD_AGPS_FILTERED is enabled, because I was under the impression that when filtered ephemerides are used, this scheduled downloads *should always* be disabled.  In your review comments today you are saying there are caveats to that, which is news to me.

Comment (tokangas) [reply subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_agps:14]: When I said that scheduled downloads should always be disabled with filtered ephemerides, I was thinking about the periodic mode, because periodic is the only mode where GNSS does scheduled downloads. In continuous tracking mode GNSS is running and receiving all the time, so there's no need for additional download logic. And in single fix mode GNSS stops after producing a single fix and doesn't do anything after that until it's started again. So, scheduled downloads only apply to the periodic mode and setting the corresponding use case bit makes no difference in other modes (it doesn't hurt either).

My point about the NRF_CLOUD_AGPS_DISABLE_SCHEDULED_DOWNLOADS option is that as an application developer I would expect scheduled downloads to be disabled if I set the option in prj.conf, but actually setting the configuration option does nothing. It exists only so that the application can check if it's enabled and act accordingly. I think it would be clearer if the application made the decision based on something else, for example based on CONFIG_NRF_CLOUD_AGPS or CONFIG_NRF_CLOUD_AGPS_FILTERED. The documentation for filtered ephemerides should say that GNSS scheduled downloads must be disabled (if periodic mode is used).

File: lib/location/method_gnss.c
Comment (hiltunent) [reply lib/location/method_gnss.c:647]: I don't see why we should want to start using GNSS in another mode so maybe I would also remove it for the reasons of clarity. I don't have a strong opinion about this.

File: applications/asset_tracker_v2/src/modules/gnss_module.c
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/gnss_module.c]: There are no immediate plans of supporting this. @jtguggedal

Comment (plskeggs) [reply applications/asset_tracker_v2/src/modules/gnss_module.c]: My reluctance to remove it is because it would be possible for a customer to change the mode, in which case this would be required if they enable filtered ephemerides. But I will remove it since you all are in agreement with each other.

File: lib/location/method_gnss.c
Comment (plskeggs) [reply lib/location/method_gnss.c:647]: I'd personally rather leave it so all the bases are covered (defensive programming), especially in case customers change the tracking mode, but I'll defer to the group.

**Review by jayteemo — APPROVED** [review]

**Review by tokangas — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst
Comment (plskeggs) [reply doc/nrf/releases/release-notes-changelog.rst:212]: It refers to including it in the request, so the set of ephemerides returned matches the desired angle.

**Pull request merged**
